### PR TITLE
[DDW-192] Improve form fields feedback UX

### PR DESCRIPTION
**Pull request by DominikGuzei**
This PR updates to `react-polymorph@0.9.7` which includes the improved `FormField` component, using `PopOver` to display errors, similar to how it was implemented in the `PasswordInput` before:

## Screenshots

CleanShot 2020-11-10 at 16 09 04

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1605021845387700)
- [ ] Test all forms within Daedalus & Storybook to check if there is anything broken / not rendering correctly.
- Validate the following inputs: 
- [x] Wallet - Send - Receiver
- [x] Wallet - Send - Amount
- [x] Wallet - Send - Confirm transaction - Password
- [ ] Wallet (Byron) - Transfer funds - Password Wallet - Transactions - Filter - Date to
- [x] Wallet - Transactions - Filter - Date from
- [x] Wallet - Transactions - Filter - Amount to
- [x] Wallet - Transactions - Filter - Amount from
- [x] Wallet - More - Settings - Name
- [x] Wallet - More - Settings - Password - Change - Current Password 
- [x] Wallet - More - Settings - Password - Change - New Password 
- [x] Wallet - More - Settings - Password - Change - Repeat Password 
- [ ] Wallet - More - Settings - Verify wallet recovery phrase - Recovery Phrase  
- [x] Wallet - More - Settings - Delete - Enter the name
- [ ] Add Wallet - Create - Wallet Name
- [ ] Add Wallet - Create - Enter Password
- [ ] Add Wallet - Create - Repeat Password
- [ ] Add Wallet - Create - Verify your recovery phrase
- [ ] Add Wallet - Restore - Recovery phrase 
- [ ] Add Wallet - Restore - Wallet Name
- [ ] Add Wallet - Restore - Enter Password
- [ ] Add Wallet - Restore - Repeat Password
- [ ] Redeem Incentivized Testnet Rewards - Recovery Phrase 
- [ ] Redeem Incentivized Testnet Rewards - Password
- [ ] Staking - Delegation Wizard - PasswordStaking - Stake Pools - Search box
---

## Testing Summary

For build **15198**

**New:**
1 - Sometimes, the red border delays to appear for the input  `Wallet - Send - Amount`. This doesn't happen all the time, but I could reproduce it 1/4 times in Windows. [Here's a recording](https://drive.google.com/file/d/1520799wHyafifP5s1SkrFEqfBorAqX8t/view?usp=sharing).
2 - There're many logs on the console now, is this expected?. [Here's a screenshot](https://drive.google.com/file/d/1AWUmt09wAR4coe08mvrNUjtexezGjwNg/view?usp=sharing).

**Fixed:**
1 -  The width of the tooltip extends nearly the same width as the app. 2) The text is not nicely aligned. 3) The tooltip position doesn't originate from the hovered element as it should. [See screenshot here](https://drive.google.com/file/d/1E2LgGDeBKn4M9-ooiNaP6l4YRZ7L65QB/view?usp=sharing)

2 - On the input `Wallet - Transaction - Filter - Date/Amount to`: when the value in "to" is smaller than in "from" an empty tooltip is displayed. Check screenshot for "[Date to](https://drive.google.com/file/d/1LkCOU83hOmg4VCccq_PLWNsfg8i4Zkac/view?usp=sharing)" and "[Amount to](https://drive.google.com/file/d/10KzMVt-S0cxtIddVSayxSyHCK1qDttRU/view?usp=sharing)"

3 - For the input `Recovery Phrase ` (On `Wallet-Create/Restore` and `Settings - Verify wallet recovery phrase`) the tooltip is not displayed until hovering the field. Check the recording [here](https://drive.google.com/file/d/17k4F3qf2NQXDBOy4anBcokwFGMX0_ewk/view?usp=sharing).

4 - This is a suggestion and needs confirmation from @nikolaglumac, as the behavior is similar in production, but is inconsistent with similar fields: 
The inputs `Redeem rewards - Recovery phrase` and `Redeem rewards - Password
` show no red border when the data is invalid/incomplete. The tooltip is displayed when hovering, so it might be unclear for the user until then. Here are a [recording](https://drive.google.com/file/d/1fTVIlF1tfj7txRW8musgT3xQoyOrMtvS/view?usp=sharing) and a [screenshot](https://drive.google.com/file/d/18mcKFeQhsaxhhAdY65Lo13UCLFH3oTtl/view?usp=sharing).

5 - The Password inputs in `Add Wallet - Restore` look smaller than before. I checked this and couldn't reproduce it in production. [Check screenshot.](https://drive.google.com/file/d/1XBSewbevGwjWauR4AGjcm6HOBGRuBeu-/view?usp=sharing)

**Rejected:**
1 - When attempted to delegate a wallet to a pool, during confirm delegation, the old style error message was displayed instead of the new tooltip style. [See here](https://drive.google.com/file/d/17_VTcvN0YJWEoLrjs781Bv4LsRhT-CtB/view?usp=sharing)

2 - On Storybook, screen `Staking - Stake Pool list` [this](https://drive.google.com/file/d/1j5rowzOgzpea-xAqCIxo7-Mx5INrmu1V/view?usp=sharing) error is displayed,

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by gnpf** [comment]
@DominikGuzei I added a list of the inputs we validated on the description of this PR. On the build 15113 these are the issues so far:

1 - On the input `Wallet - Transaction - Filter - Date/Amount to`: when the value in "to" is smaller than in "from" an empty tooltip is displayed. Check screenshot for "[Date to](https://drive.google.com/file/d/1LkCOU83hOmg4VCccq_PLWNsfg8i4Zkac/view?usp=sharing)" and "[Amount to](https://drive.google.com/file/d/10KzMVt-S0cxtIddVSayxSyHCK1qDttRU/view?usp=sharing)"

2 - For the input `Recovery Phrase ` (On `Wallet-Create/Restore` and `Settings - Verify wallet recovery phrase`) the tooltip is not displayed until hovering the field. Check the recording [here](https://drive.google.com/file/d/17k4F3qf2NQXDBOy4anBcokwFGMX0_ewk/view?usp=sharing).

3 - This is a suggestion and needs confirmation from @nikolaglumac, as the behavior is similar in production, but is inconsistent with similar fields: 
The inputs `Redeem rewards - Recovery phrase` and `Redeem rewards - Password
` show no red border when the data is invalid/incomplete. The tooltip is displayed when hovering, so it might be unclear for the user until then. Here are a [recording](https://drive.google.com/file/d/1fTVIlF1tfj7txRW8musgT3xQoyOrMtvS/view?usp=sharing) and a [screenshot](https://drive.google.com/file/d/18mcKFeQhsaxhhAdY65Lo13UCLFH3oTtl/view?usp=sharing).

4 - On Storybook, screen `Staking - Stake Pool list` [this](https://drive.google.com/file/d/1j5rowzOgzpea-xAqCIxo7-Mx5INrmu1V/view?usp=sharing) error is displayed,

**Comment by mioriohk** [comment]
Hi @DominikGuzei, hope you are well. Here are the issues on build 15115. Thank you:

**New**
1 - When attempted to delegate a wallet to a pool, the old style error message was displayed instead of the new tooltip style. [See here](https://drive.google.com/file/d/17_VTcvN0YJWEoLrjs781Bv4LsRhT-CtB/view?usp=sharing)

**Pending**
1 - On the input `Wallet - Transaction - Filter - Date/Amount to`: when the value in "to" is smaller than in "from" an empty tooltip is displayed. Check screenshot for "[Date to](https://drive.google.com/file/d/1LkCOU83hOmg4VCccq_PLWNsfg8i4Zkac/view?usp=sharing)" and "[Amount to](https://drive.google.com/file/d/10KzMVt-S0cxtIddVSayxSyHCK1qDttRU/view?usp=sharing)"

2 - For the input `Recovery Phrase ` (On `Wallet-Create/Restore` and `Settings - Verify wallet recovery phrase`) the tooltip is not displayed until hovering the field. Check the recording [here](https://drive.google.com/file/d/17k4F3qf2NQXDBOy4anBcokwFGMX0_ewk/view?usp=sharing).

3 - This is a suggestion and needs confirmation from @nikolaglumac, as the behavior is similar in production, but is inconsistent with similar fields: 
The inputs `Redeem rewards - Recovery phrase` and `Redeem rewards - Password
` show no red border when the data is invalid/incomplete. The tooltip is displayed when hovering, so it might be unclear for the user until then. Here are a [recording](https://drive.google.com/file/d/1fTVIlF1tfj7txRW8musgT3xQoyOrMtvS/view?usp=sharing) and a [screenshot](https://drive.google.com/file/d/18mcKFeQhsaxhhAdY65Lo13UCLFH3oTtl/view?usp=sharing).

4 - On Storybook, screen `Staking - Stake Pool list` [this](https://drive.google.com/file/d/1j5rowzOgzpea-xAqCIxo7-Mx5INrmu1V/view?usp=sharing) error is displayed,

**Comment by nikolaglumac** [comment]
> 3 - This is a suggestion and needs confirmation from @nikolaglumac, as the behavior is similar in production, but is inconsistent with similar fields:
> The inputs `Redeem rewards - Recovery phrase` and `Redeem rewards - Password ` show no red border when the data is invalid/incomplete. The tooltip is displayed when hovering, so it might be unclear for the user until then. Here are a [recording](https://drive.google.com/file/d/1fTVIlF1tfj7txRW8musgT3xQoyOrMtvS/view?usp=sharing) and a [screenshot](https://drive.google.com/file/d/18mcKFeQhsaxhhAdY65Lo13UCLFH3oTtl/view?usp=sharing).

@gnpf @DominikGuzei the correct behavior would be that all fields in error state have red border - tooltip is shown on hover or on focus on the field.

cc @daniloprates @a-rukin

**Comment by gnpf** [comment]
@DominikGuzei For the build 15115, here's a new summary:

**New**
1 - The Password inputs in `Add Wallet - Restore` look smaller than before. I checked this and couldn't reproduce it in production.

**Pending**
1 - When attempted to delegate a wallet to a pool, during confirm delegation, the old style error message was displayed instead of the new tooltip style. [See here](https://drive.google.com/file/d/17_VTcvN0YJWEoLrjs781Bv4LsRhT-CtB/view?usp=sharing)

2 - On the input `Wallet - Transaction - Filter - Date/Amount to`: when the value in "to" is smaller than in "from" an empty tooltip is displayed. Check screenshot for "[Date to](https://drive.google.com/file/d/1LkCOU83hOmg4VCccq_PLWNsfg8i4Zkac/view?usp=sharing)" and "[Amount to](https://drive.google.com/file/d/10KzMVt-S0cxtIddVSayxSyHCK1qDttRU/view?usp=sharing)"

3 - For the input `Recovery Phrase ` (On `Wallet-Create/Restore` and `Settings - Verify wallet recovery phrase`) the tooltip is not displayed until hovering the field. Check the recording [here](https://drive.google.com/file/d/17k4F3qf2NQXDBOy4anBcokwFGMX0_ewk/view?usp=sharing).

4 - This is a suggestion and needs confirmation from @nikolaglumac, as the behavior is similar in production, but is inconsistent with similar fields: 
The inputs `Redeem rewards - Recovery phrase` and `Redeem rewards - Password
` show no red border when the data is invalid/incomplete. The tooltip is displayed when hovering, so it might be unclear for the user until then. Here are a [recording](https://drive.google.com/file/d/1fTVIlF1tfj7txRW8musgT3xQoyOrMtvS/view?usp=sharing) and a [screenshot](https://drive.google.com/file/d/18mcKFeQhsaxhhAdY65Lo13UCLFH3oTtl/view?usp=sharing).

5 - On Storybook, screen `Staking - Stake Pool list` [this](https://drive.google.com/file/d/1j5rowzOgzpea-xAqCIxo7-Mx5INrmu1V/view?usp=sharing) error is displayed,

**Comment by DominikGuzei** [comment]
@nikolaglumac @darko-mijic @a-rukin please check issue 2 reported by @gnpf ☝️ is that OK? This PR is only concerned about updating to the improved `FormField` component, it doesn't handle high-level errors that are not related to a specific form field.

**Comment by DominikGuzei** [comment]
@gnpf regarding PENDING 2: how did you enter the invalid dates? The date pickers didn't let me pick the invalid combination

**Comment by a-rukin** [comment]
> @nikolaglumac @darko-mijic @a-rukin please check issue PENDING 1 reported by @gnpf ☝️ is that OK? 

hm... it looks like error connected to the password field, no?

**Comment by gnpf** [comment]
> @gnpf regarding PENDING 2: how did you enter the invalid dates? The date pickers didn't let me pick the invalid combination

@DominikGuzei To reproduce this issue, follow these steps:

1 - Select "All transactions"
2 - Select on "to" any date allowed
3 - Select on "from" a posterior date

Check [this](https://drive.google.com/file/d/1pPKxsSW-eCIYsc0r-kUjDrBS2sThOLBH/view?usp=sharing) recording (it's an unrelated build)

**Comment by DominikGuzei** [comment]
@a-rukin yes and no … it is related to password input but it's not a regression in the scope of this PR, since this error is simply not passed to the password input form field itself but displayed as an "server error". But i will check how easy a fix to this issue would be 👍

**Comment by nikolaglumac** [comment]
> 1 - When attempted to delegate a wallet to a pool, during confirm delegation, the old style error message was displayed instead of the new tooltip style. See here

☝️ this is a server side error and it looks as expected. We do not expect this error to be shown in a form of a tooltip on the field at this moment. We will handle this in a separate card.

cc @DominikGuzei @gnpf

**Comment by DominikGuzei** [comment]
@gnpf @nikolaglumac all reported QA issues that are relevant to this PR should be fixed now:

New
1 - The Password inputs in Add Wallet - Restore look smaller than before -> **FIXED**

Pending
1 - When attempted to delegate a wallet to a pool, during confirm delegation, the old style error message was displayed instead of the new tooltip style. -> **EXPECTED, WILL NOT FIX**

2 - On the input Wallet - Transaction - Filter - Date/Amount to: when the value in "to" is smaller than in "from" an empty tooltip is displayed. Check screenshot for "Date to" and "Amount to" -> **FIXED**

3 - For the input Recovery Phrase  (On Wallet-Create/Restore and Settings - Verify wallet recovery phrase) the tooltip is not displayed until hovering the field. Check the recording here. -> **FIXED**

4 - This is a suggestion and needs confirmation from @nikolaglumac, as the behavior is similar in production, but is inconsistent with similar fields:
The inputs Redeem rewards - Recovery phrase and Redeem rewards - Password show no red border when the data is invalid/incomplete. The tooltip is displayed when hovering, so it might be unclear for the user until then. Here are a recording and a screenshot. -> **FIXED**

5 - On Storybook, screen Staking - Stake Pool list this error is displayed, > **NOT FIXED, UNRELATED**

**Comment by gnpf** [comment]
@DominikGuzei Great job, most of the issues are fixed, but there is one that I can still reproduce. For the build 15115, here's a new summary:

**Reincident**

1 - On the input `Wallet - Transaction - Filter - Date/Amount to`: when the value in "to" is smaller than in "from" an empty tooltip is displayed. Check screenshot for "[Date to](https://drive.google.com/file/d/1LkCOU83hOmg4VCccq_PLWNsfg8i4Zkac/view?usp=sharing)" and "[Amount to](https://drive.google.com/file/d/10KzMVt-S0cxtIddVSayxSyHCK1qDttRU/view?usp=sharing)"
New screenshot [here](https://drive.google.com/file/d/1sFedC1wjDPbPc1AB7XBi_-Rmh2xsWOoi/view?usp=sharing).

**Fixed**
1 - The Password inputs in `Add Wallet - Restore` look smaller than before. I checked this and couldn't reproduce it in production. [Check screenshot.](https://drive.google.com/file/d/1XBSewbevGwjWauR4AGjcm6HOBGRuBeu-/view?usp=sharing)

2 - For the input `Recovery Phrase ` (On `Wallet-Create/Restore` and `Settings - Verify wallet recovery phrase`) the tooltip is not displayed until hovering the field. Check the recording [here](https://drive.google.com/file/d/17k4F3qf2NQXDBOy4anBcokwFGMX0_ewk/view?usp=sharing).

3 - This is a suggestion and needs confirmation from @nikolaglumac, as the behavior is similar in production, but is inconsistent with similar fields: 
The inputs `Redeem rewards - Recovery phrase` and `Redeem rewards - Password
` show no red border when the data is invalid/incomplete. The tooltip is displayed when hovering, so it might be unclear for the user until then. Here are a [recording](https://drive.google.com/file/d/1fTVIlF1tfj7txRW8musgT3xQoyOrMtvS/view?usp=sharing) and a [screenshot](https://drive.google.com/file/d/18mcKFeQhsaxhhAdY65Lo13UCLFH3oTtl/view?usp=sharing).

**Comment by DominikGuzei** [comment]
@gnpf the last issue with the tx filter should be fixed now 👍

File: yarn.lock [hunk 1]
```diff
@@ -11967,9 +11967,10 @@ react-modal@3.1.12:
     prop-types "^15.5.10"
     warning "^3.0.0"
 
-react-polymorph@0.9.6-rc.3:
-  version "0.9.6-rc.3"
-  resolved "https://registry.yarnpkg.com/react-polymorph/-/react-polymorph-0.9.6-rc.3.tgz#dceb56aa8e7d83dc8251d7459257d0934c45b104"
+react-polymorph@0.9.7-rc.4:
+  version "0.9.7-rc.4"
+  resolved "https://registry.yarnpkg.com/react-polymorph/-/react-polymorph-0.9.7-rc.4.tgz#77117f537c5bd60c68d7cd8cce2f79e0d13e94d1"
+  integrity sha512-LP2XijEr20B8CCgXozCNGN1JxUE38R+sRvqArssbUxxCV0V7WS7TzSLCWJgjU07P8xDlBejymqMfBZVDASh4Og==
```
Comment (nikolaglumac) [inline yarn.lock:11973]: @DominikGuzei please remove this line 🙏

Comment (DominikGuzei) [reply yarn.lock]: done 👍

**Comment by mioriohk** [comment]
Hi @DominikGuzei, as discussed on Slack

For build 15197:

**New issue**
On the tooltip for "Paper Wallet" in "More" tab.
1) The width of the tooltip extends nearly the same width as the app. 2) The text is not nicely aligned. 3) The tooltip position doesn't originate from the hovered element as it should. [See screenshot here](https://drive.google.com/file/d/1E2LgGDeBKn4M9-ooiNaP6l4YRZ7L65QB/view?usp=sharing)

**Comment by DominikGuzei** [comment]
@mioriohk the paper wallet tooltip should look better now 👍

**Review by mioriohk — APPROVED** [review]
Looks good to me @DominikGuzei . Great job! 👍

File: source/renderer/app/components/wallet/settings/WalletRecoveryPhraseVerificationWidget.scss [hunk 2]
```diff
@@ -105,4 +105,5 @@
 
 .paperWalletTooltip {
   line-height: 22px;
+  width: 40em;
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/settings/WalletRecoveryPhraseVerificationWidget.scss:108]: @DominikGuzei can you please rather use "px" here? We are not using "em" anywhere else...

**Comment by DominikGuzei** [comment]
@gnpf great catches! 👍 
1 - yeah the form field always delays the error state to 1 second after the last change of the error message (so I guess in this case the final error message returns after calculating the fees and then it waits another 1 second before displaying it) … which is not optimal in this very case but required to make the typing UX better (e.g: that the error is only shown after a short delay while typing) - but now that im thinking of it - the problem might be better solved by using the last input change as the start of the delay (then it would work for the typing case but would show immediately in the calculation case)

File: source/renderer/app/components/wallet/settings/WalletRecoveryPhraseVerificationWidget.scss (hunk 2 above)
Comment (DominikGuzei) [reply source/renderer/app/components/wallet/settings/WalletRecoveryPhraseVerificationWidget.scss]: done 👍

**Comment by DominikGuzei** [comment]
@nikolaglumac @gnpf I removed the form field error delay as I figured that it's not the right place to make this decision - the error delay should be provided by the form component (by debouncing the input change before passing on the error). The form field component is not really in the position of handling this kind of UX automatically. This means the UX will be more similar to what we already had in Daedalus (but with pop overs which fade-in to display the errors)

**Comment by nikolaglumac** [comment]
@a-rukin please review this one 🙏

**Review by gnpf — APPROVED** [review]

**Comment by DominikGuzei** [comment]
@a-rukin none of these points are related to this PR:
1: I don't know why it shows this pending tx error, but it surely is not an issue introduced in this PR (can you try to reproduce on `develop`?)
2 + 3: Yes, that's expected as it is a server error which is not passed to the password form field
4: No agreement yet, although I wouldn't say this is a blocker for this PR
5: I am pretty sure that this is not related (can you try to reproduce on `develop`?)

**Comment by a-rukin** [comment]
> 2 + 3: Yes, that's expected as it is a server error which is not passed to the password form field

would it be possible to handle them in the same tooltip way? (maybe in not this PR, but in future)

**Review by a-rukin — APPROVED** [review]

**Comment by DominikGuzei** [comment]
@a-rukin i'm not sure, we need to check that - but i guess it will be possible 👍

**Review by nikolaglumac — APPROVED** [review]
Great work @DominikGuzei 💯

**Pull request merged**